File: Quick-Settings/app/src/main/java/com/google/android_quick_settings/MainActivity.java
// Copyright 2016 Google Inc.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android_quick_settings;

import android.app.StatusBarManager;
import android.content.ComponentName;
import android.os.Build.VERSION;
import android.os.Bundle;
import android.widget.Toast;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatButton;
import androidx.core.graphics.drawable.IconCompat;
import com.google.common.util.concurrent.MoreExecutors;
import java.util.logging.Level;
import java.util.logging.Logger;

public class MainActivity extends AppCompatActivity {

  @RequiresApi(api = 33)
  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    StatusBarManager statusBarService = this.getSystemService(StatusBarManager.class);
    ComponentName componentName = new ComponentName(
        this.getApplicationContext(),
        QuickSettingsService.class.getName());
    AppCompatButton btn = findViewById(R.id.add_tile_btn);
    IconCompat icon =
        IconCompat.createWithResource(getApplicationContext(),
            R.drawable.ic_android_black_24dp);

    btn.setOnClickListener(view -> {
      if (VERSION.SDK_INT == 33) {
        statusBarService.requestAddTileService(
            componentName, "Quick Settings", icon.toIcon(MainActivity.this),
            MoreExecutors.directExecutor(), integer -> {
               setResult(integer);
               finish();
            });
      } else {
        Logger.getLogger(MainActivity.class.getName()).log(
            Level.INFO, "Request to add tile for user is not supported");
      }
    });
  }
}

File: AppWidget/glance-widget/src/main/java/com/example/android/glancewidget/MainActivity.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.glancewidget

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Surface
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import com.example.android.glancewidget.ui.theme.AppWidgetTheme

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            AppWidgetTheme {
                // A surface container using the 'background' color from the theme
                Surface(color = MaterialTheme.colors.background) {
                    Content()
                }
            }
        }
    }
}

@Composable
fun Content() {
    Box(modifier = Modifier.fillMaxSize()) {
        Text(
            text = stringResource(R.string.activity_for_glance_widgets),
            modifier = Modifier.align(Alignment.Center)
        )
    }
}

@Preview(showBackground = true)
@Composable
fun ContentPreview() {
    AppWidgetTheme {
        Content()
    }
}
